### PR TITLE
Start adding server response support

### PR DESCRIPTION
**Pull request by BrennanConroy**
+ New Handshake response protocol update

/cc @anurse 

Definitely needs some work:
* Locking around `mPendingCalls`
* ~~Handle error from `CompletionMessage`~~
* ~~InvocationId not always `"1"`~~

### CODE REVIEW

File: src/signalrlib/HubConnection.cpp [hunk 1]
```diff
@@ -9,21 +9,35 @@ namespace signalr
...
         mTransport->OnReceived([&](const utility::string_t& message)
         {
+            // trim '\x1e' then parse
             auto value = web::json::value::parse(message.substr(0, message.size() - 1));
+            if (!mHandshakeReceived)
+            {
+                // A failed handshake or not a handshake response
+                if (value.has_field(L"error") || value.size() != 0)
+                {
+                    // TODO: Logging
```
Comment (anurse) [inline src/signalrlib/HubConnection.cpp]: Is there an issue tracking logging?

File: src/signalrlib/HubConnection.cpp [hunk 2]
```diff
@@ -48,8 +62,23 @@ namespace signalr
                 // TODO
                 break;
             case MessageType::Completion:
-                // TODO
+            {
+                auto id = value[L"invocationId"];
+                auto it = mPendingCalls.find(id.as_string());
+                auto invocationRequest = (*it).second;
```
Comment (anurse) [inline src/signalrlib/HubConnection.cpp]: `->` does this, right? `it->second`

File: src/signalrlib/HubConnection.cpp [hunk 3]
```diff
@@ -9,21 +9,35 @@ namespace signalr
...
     HubConnection::HubConnection(const utility::string_t& url, Transport transport)
+        : mHandshakeReceived(false)
     {
         mUrl = url;
         mTransport = new WebSocketsTransport(mUrl);
         mTransport->OnReceived([&](const utility::string_t& message)
         {
+            // trim '\x1e' then parse
             auto value = web::json::value::parse(message.substr(0, message.size() - 1));
+            if (!mHandshakeReceived)
```
Comment (anurse) [inline src/signalrlib/HubConnection.cpp]: We're not guaranteed that `OnReceived` will be called serially. We should lock.

File: src/signalrlib/HubConnection.cpp [hunk 4]
```diff
@@ -48,8 +62,23 @@ namespace signalr
...
+
+                if (value.has_field(L"error"))
+                {
+                    auto error = value[L"error"].as_string();
+                    invocationRequest.set_exception(std::runtime_error(utility::conversions::to_utf8string(error)));
+                    break;
+                }
+
+                auto val = value[L"result"];
+                invocationRequest.set(val.serialize());
```
Comment (anurse) [inline src/signalrlib/HubConnection.cpp]: I think it's clearer to put this in an `else` clause and just have the `break` outside the whole `if`.

File: src/signalrlib/HubConnection.cpp (hunk 1 above)
Comment (BrennanConroy) [reply src/signalrlib/HubConnection.cpp]: Kind of https://github.com/aspnet/SignalR-Client-Cpp/issues/3

File: src/signalrlib/HubConnection.cpp [hunk 5]
```diff
@@ -9,66 +9,116 @@ namespace signalr
...
-                _ASSERT(args.is_array());
-                auto m = method.serialize();
-                // serializing a string creates "method", strip quotes, also figure out a better way to do this
-                auto handler = mHandlers.find(m.substr(1, m.size() - 2));
-                if (handler != mHandlers.end())
+                // TODO: message can have multiple messages
+                auto value = web::json::value::parse(message.substr(0, message.size() - 1));
+                {
+                    std::lock_guard<std::mutex> lock(mLock);
+                    // trim '\x1e' then parse
```
Comment (anurse) [inline src/signalrlib/HubConnection.cpp]: Comment is in the wrong place :)

File: src/signalrlib/HubConnection.cpp (hunk 1 above)
Comment (anurse) [reply src/signalrlib/HubConnection.cpp]: Good enough for me

**Pull request merged**